server/re/src/msi_genquery2.cpp

```
@@ -0,0 +1,315 @@
+/// # the resultset has been reached.
+/// END_OF_RESULTSET = -408000;
+///
+/// iterating_over_genquery2_results()
+/// {
+///     # Execute a query. The results are stored in the Rule Engine Plugin.
+///     msi_genquery2_execute(*handle, "select COLL_NAME, DATA_NAME order by DATA_NAME desc limit 1");
+///
+///     # Iterate over the results.
+///     while (true) {
+///         *ec = errorcode(msi_genquery2_next_row(*handle));
+///
+///         if (*ec < 0) {
+///             if (END_OF_RESULTSET == *ec) {
+///                 break;
+///             }
+///
+///             failmsg(*ec, "Unexpected error while iterating over GenQuery2 resultset.");
+///         }
+///
+///         msi_genquery2_column(*handle, '0', *coll_name); # Copy the COLL_NAME into *coll_name.
+///         msi_genquery2_column(*handle, '1', *data_name); # Copy the DATA_NAME into *data_name.
+///         writeLine("stdout", "logical path => [*coll_name/*data_name]");
+///     }
+///
+///     # Free any resources used. This is handled for you when the agent is shut down as well.
+///     msi_genquery2_free(*handle);
+/// }
+/// \endcode
+///
+/// \since 4.3.2
+auto msi_genquery2_next_row(MsParam* _handle, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_handle->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->inOutStruct);
+
+    try {
+        const auto ctx_handle_index = std::stoll(static_cast<char*>(_handle->inOutStruct));
+
+        if (ctx_handle_index < 0) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        auto iter = gq2_context.find(ctx_handle_index);
+
+        if (std::end(gq2_context) == iter) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+        }
```

Should we return an error here? Or is it intentionally let go because the `std::exception` is caught below?

---

Yes. Good catch.

---

Done